File: ice0208/two-pointers/1679.py
class Solution:
    def maxOperations(self, nums: list[int], k: int) -> int:
        # 요약: 정렬 후 투 포인터

        # nums를 정렬합니다.
        nums.sort()

        # left와 right포인터를 초기 세팅합니다.
        left, right = 0, len(nums) - 1

        # left와 right 사이의 범위를 서서히 줄입니다.
        # 범위를 줄이면서 left와 right 대상의 합이 k이면 count를 셉니다.
        count = 0
        while left < right:
            cur_sum = nums[left] + nums[right]

            if cur_sum == k:
                count += 1
                left += 1
                right -= 1

            elif cur_sum > k:
                right -= 1
            else:
                left += 1

        return count


if __name__ == "__main__":
    print(Solution().maxOperations(nums=[1, 2, 3, 4], k=5))
    print(Solution().maxOperations(nums=[3, 1, 3, 4, 3], k=6))

File: moonpoison/two_pointers/1679.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) 
    {
        int ans=0;
        int p1=0, p2=nums.size()-1;
        sort(nums.begin(), nums.end());
        while(p1<p2)
        {
            int sum=nums[p1]+nums[p2];
            if(sum==k)
            {
                ans++;
                p1++; p2--;
            }
            else if(sum<k) p1++;
            else p2--;
        }
        return ans;
    }
};